### ISSUE TITLE
Applying SHAP to pycox models

### ISSUE THREAD

**Issue by VanishingRasengan** [havakv/pycox#126]
Dear @havakv Sir,

first of all I want to say thank you for maintaining this amazing python package for survival analysis!

However, I am trying to apply SHAP (https://github.com/slundberg/shap), a tool for explainable AI, on pycox but it somehow does not seems to be compatible. 

I followed the same steps like here (https://nbviewer.org/github/havakv/pycox/blob/master/examples/cox-ph.ipynb?ref=https://githubhelp.com) and put this into the end:

`import shap

explainer = shap.Explainer(model, x_train)
shaps = explainer(X_test)`

The Error message is:

`The passed model is not callable and cannot be analyzed directly with the given masker! Model: <pycox.models.cox.CoxPH object at 0x7f88195c3d90>`

Is pycox not compatible with SHAP or did I apply the "Explainer" method wrongly?

Kind regards,
Muralee

**Comment by havakv** [comment]
Hi, so we haven't made any effort to make pycox compatible with SHAP. I can see that there is some preliminary support for pytorch in SHAP, but I'm really not familiar with it. You could try to make it work by following som pytorch examples in SHAP and try to replicate that with pycox if you feel like it.

I guess full support would be nice, if someone were to look into it

**Comment by teituroli** [comment]
> Dear @havakv Sir,
> 
> first of all I want to say thank you for maintaining this amazing python package for survival analysis!
> 
> However, I am trying to apply SHAP (https://github.com/slundberg/shap), a tool for explainable AI, on pycox but it somehow does not seems to be compatible.
> 
> I followed the same steps like here (https://nbviewer.org/github/havakv/pycox/blob/master/examples/cox-ph.ipynb?ref=https://githubhelp.com) and put this into the end:
> 
> ```
> import shap
> 
> explainer = shap.Explainer(model, x_train)
> shaps = explainer(X_test)
> ```
> 
> The Error message is:
> 
> `The passed model is not callable and cannot be analyzed directly with the given masker! Model: <pycox.models.cox.CoxPH object at 0x7f88195c3d90>`
> 
> Is pycox not compatible with SHAP or did I apply the "Explainer" method wrongly?
> 
> Kind regards, Muralee

Hi, Is there any chance you have found a solution? 

All the bests,
T

**Comment by thecml** [comment]
I haven't tested myself with Pycox but found good results with scikit-survival and SHAP using the predict-trick, so my guess is that the same works:

```
import shap

explainer = shap.Explainer(model.predict, X_train, feature_names=feature_names)
shaps = explainer(X_test)
```

**Comment by hellorp1990** [comment]
@VanishingRasengan @teituroli Hi I am having a similar issue with SHAP. Do you have a solution on how to use SHAP for pycox?

**Comment by teituroli** [comment]
Hi @hellorp1990,
I did not spend too much time trying to find a solution. 
I wrote my own loss function based on the Cox Proportional hazards model. 
You can find the very messy code [here](https://github.com/teituroli/Thesis/blob/main/Python_Code/Combined_large_network.ipynb). 
Feel free to ask any questions.

**Comment by LuisPerezLombardia** [comment]
Hi, 
This seems to be working for me for the linear synthetic dataset generated in:

**Katzman, J.L., Shaham, U., Cloninger, A. et al. (2018).** [*DeepSurv: personalized treatment recommender system using a Cox proportional hazards deep neural network.*](https://doi.org/10.1186/s12874-018-0482-1) BMC Med Res Methodol, 18, 24.**


Here:

h(x) =  x_0 + 2*x_1 


```python
import torch
import torchtuples as tt
import numpy as np
import shap
from pycox.models import CoxPH
from torch import nn

# Defining the network architecture
net = tt.practical.MLPVanilla(in_features, num_nodes, out_features, batch_norm,
                              dropout, activation=nn.SELU, output_bias=output_bias)

# Setting up the optimizer
optimizer = tt.optim.SGD(lr=0.0002922, momentum=0.906, weight_decay=0.000399)
model_0a = CoxPH(net, optimizer)

# Training the model
epochs = 100
callbacks = [tt.callbacks.EarlyStopping()]
batch_size = 50
log = model_0a.fit(x_train, y_train, batch_size, epochs, callbacks, val_data=val)
_ = log.plot()

# Predicting risk scores
def predict_risk(x):
    if not isinstance(x, torch.Tensor):
        x = torch.tensor(x, dtype=torch.float32)
    with torch.no_grad():
        return model_0a.predict(x).cpu().numpy().astype(np.float32)

print(predict_risk(x_train[:10]).dtype)

# SHAP evaluation
col = ['x0', 'x1', 'x2', 'x3','x4', 'x5', 'x6', 'x7', 'x8', 'x9']
explainer = shap.Explainer(predict_risk, x_train[:1000], feature_names=col)
shap_values = explainer(x_train[:1000])

# Plotting the SHAP values
shap.summary_plot(shap_values, x_train[:1000])

```
Here
![SHAP linear](https://github.com/havakv/pycox/assets/84937565/67ffe31f-69cb-4add-b674-89677c40e992)

**Comment by teituroli** [comment]
@LuisPerezLombardia, wonderful that you found a solution!

**Comment by Thomson-Cui** [comment]
@LuisPerezLombardia, excellent work!
